### PR TITLE
Core: allow loading worlds from zip modules

### PR DESCRIPTION
**Pull request by Berserker66**
Core: allow loading worlds from zip modules
RoR2: make it zipimport compatible (remove relative imports beyond local top-level)
^ my test world

Should be mostly functional as-is, many worlds won't function as-is, however. Worlds like Factorio or Subnautica would need to be taught to load those files from within a zip, many worlds import relatively beyond their would-be zip scope. 

Some todos:
Should figure out how we handle requirements.txt, even if that handling is just raising an error that it's not supported (yet?).
Should provide a way in AP to create a world zip easily and throw meaningful exceptions if problems are found ahead of time. (zipfile.PyZipFile is likely a good starting point.)

### CODE REVIEW

**Comment by alwaysintreble** [comment]
`FileNotFoundError: [WinError 3] The system cannot find the path specified: "...\\Archipelago\\worlds\\ror2.apworld\\ror2\\docs'`

**Comment by Ijwu** [comment]
What's the justification or intent with this?

**Comment by black-sliver** [comment]
I'm wondering if we should just merge this and note it as experimental feature somewhere in docs, so people can play around with it. I'm afraid that this may get left out when someone substantially modifies AutoWorld if this is not merged (i.e. when optimizing or building a cache or adding metadata).
Should maybe also have a test that zips up Generic (or any other world) and tries to load from zip - maybe even just an pytest.raises for `raise RuntimeError(f"""Game {dct["game"]} already registered.""")`.

**Comment by Jarno458** [comment]
How is this regarding security? wont the world zip contain .py files? like wont this be loading untrusted python files?

**Comment by Berserker66** [comment]
> How is this regarding security? wont the world zip contain .py files? like wont this be loading untrusted python files?

the same it currently is with frozen distribution. Files being in a zip or a folder does not change much security wise.

**Comment by Berserker66** [comment]
So, this should be fully functional but missing some bells and whistles. Though it might be a in a good place to include, find a usecase in the next version and improve tooling from there.

For example, had this been included in 0.3.4 The Witness could have supplied an apworld to fix a logic bug that was found shortly after release and we could have used that to figure out stumbling blocks and tools needed. So I'm hoping for a similar opportunity in the future.

Independent of that, potentially as a follow-up PR; any interest in a World flag to indicate distribution as .apworld so that setup.py takes care of packaging and the world is as .apworld in the .exe?

File: worlds/AutoWorld.py [hunk 1]
```diff
@@ -170,9 +179,10 @@ class World(metaclass=AutoWorldRegister):
     item_names: Set[str]  # set of all potential item names
     location_names: Set[str]  # set of all potential location names
 
-    web: WebWorld = WebWorld()
+    is_zip: bool  # was loaded from a .apworld ?
```
Comment (black-sliver) [inline worlds/AutoWorld.py:182]: this should probably be read-only/a property?

Comment (Berserker66) [reply worlds/AutoWorld.py:182]: I actually had it as property first, but the way the World is documented below it made it look like you would be expected/intended to overwrite the method.

**Comment by black-sliver** [comment]
Tested and looks good (besides missing upgrade path in installer and meta data, but that's not required for testing). Waiting for  #902 and #903 to be reviewed and merged and then we can just have this live in main to play around.

**Review by black-sliver — APPROVED** [review]
Tested with RoR2. Seems to behave exactly as expected.

**Pull request merged**